util/network/src/client_version.rs

```diff
@@ -0,0 +1,516 @@
+			identity: identity,
+			semver: semver,
+			os: os,
+			compiler: compiler,
+
+			can_handle_large_requests: can_handle_large_requests,
+		}
+	}
+
+	fn name(&self) -> &str {
+		self.name.as_str()
+	}
+
+	fn identity(&self) -> Option<&str> {
+		self.identity.as_ref().map(String::as_str)
+	}
+
+	fn semver(&self) -> &Version {
+		&self.semver
+	}
+
+	fn os(&self) -> &str {
+		self.os.as_str()
+	}
+
+	fn compiler(&self) -> &str {
+		self.compiler.as_str()
+	}
+
+	fn can_handle_large_requests(&self) -> bool {
+		self.can_handle_large_requests
+	}
+}
+
+/// Enum describing the version of the software running on a peer.
+#[derive(Clone,Debug,Eq,PartialEq,Serialize)]
+pub enum ClientVersion {
+	/// The peer runs software from parity and the string format is known
+	ParityClient(
+		/// The actual information fields: name, version, os, ...
+		ParityClientData
+	),
+	/// The string ID is recognized as Parity but the overall format
+	/// could not be parsed
+	ParityUnknownFormat(String),
+	/// Other software vendors than Parity
+	Other(String),
+}
+
+impl Default for ClientVersion {
+	fn default() -> Self {
+		ClientVersion::Other("".to_owned())
+	}
+}
+
+/// Provide information about what a particular version of a
+/// peer software can do
+pub trait ClientCapabilities {
+	/// Parity versions before PARITY_CLIENT_LARGE_REQUESTS_VERSION would not
+	/// check the accumulated size of a packet when building a response to a
+	/// GET_BLOCK_BODIES request. If the packet was larger than a given limit,
+	/// instead of sending fewer blocks no packet would get sent at all. Query
+	/// if this version can handle requests for a large number of block bodies.
+	fn can_handle_large_requests(&self) -> bool;
+
+	/// Service transactions are specific to parity. Query if this version
+	/// accepts them.
+	fn accepts_service_transaction(&self) -> bool;
+}
+
+impl ClientCapabilities for ClientVersion {
+	fn can_handle_large_requests(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => data.can_handle_large_requests(),
+			ClientVersion::ParityUnknownFormat(_) => false, // Play it safe
+			ClientVersion::Other(_) => true // As far as we know
+		}
+	}
+
+	fn accepts_service_transaction(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(_) => true,
+			ClientVersion::ParityUnknownFormat(_) => true,
+			ClientVersion::Other(_) => false
+		}
+	}
+
+}
+
+fn is_parity(client_id: &str) -> bool {
+	client_id.starts_with(LEGACY_CLIENT_ID_PREFIX) || client_id.starts_with(PARITY_CLIENT_ID_PREFIX)
+}
+
+/// Parse known parity formats. Recognizes either a short format with four fields
+/// or a long format which includes the same fields and an identity one.
+fn parse_parity_format(client_version: &str) -> Result<ParityClientData, ()> {
+	let tokens: Vec<&str> = client_version.split("/").collect();
+
+	if tokens.len() < PARITY_ID_STRING_MINIMUM_LENGTH {
+		return Err(())
+	}
+
+	let name = tokens[0];
+
+	let identity = if tokens.len() - 3 > 1 {
```

Comment:
This feels simpler:
```suggestion
	let identity = if tokens.len() > PARITY_ID_STRING_MINIMUM_LENGTH {
```

Reply:
I actually prefer the first alternative for consistency, because in other places we can find `tokens.len() - n`, so it fits better with the rest of the code. The semantics is more clear to me: if the position where identity should be is greater than one, then actually find identity. What I feel really unhappy about is using hardwired numbers. I considered using constants or macros but could not come up with anything beautiful.